# Week 3: Supervised Learning — Classification Model

**Intern:** Fiza Zahid Ali  
**Discipline:** Bachelor of Science in Artificial Intelligence (BSAI)  
**Organization:** Zynxis  
**Task:** Supervised Learning — Classification
## 1. Project Objective

The objective of this task is to build supervised machine learning classification models to predict whether a Titanic passenger survived based on passenger-related features.

Three classification algorithms will be trained and compared:

1. Logistic Regression
2. Decision Tree Classifier
3. Random Forest Classifier

The models will be evaluated using:

- Accuracy
- Precision
- Recall
- F1 Score

The best-performing model will then be selected based on the evaluation results.

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

In [2]:
df = pd.read_csv("Cleaned_Titanic.csv")

df.head()
df.shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 825 entries, 0 to 824
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Survived       825 non-null    int64  
 1   Pclass         825 non-null    int64  
 2   Sex            825 non-null    int64  
 3   Age            825 non-null    float64
 4   SibSp          825 non-null    int64  
 5   Parch          825 non-null    int64  
 6   Fare           825 non-null    float64
 7   Embarked       825 non-null    object 
 8   FamilySize     825 non-null    int64  
 9   FarePerPerson  825 non-null    float64
 10  IsAlone        825 non-null    int64  
 11  AgeGroup       825 non-null    int64  
dtypes: float64(3), int64(8), object(1)
memory usage: 77.5+ KB


In [4]:
df.isnull().sum()


,0
Survived,0
Pclass,0
Sex,0
Age,0
SibSp,0
Parch,0
Fare,0
Embarked,0
FamilySize,0
FarePerPerson,0


In [5]:
df.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked', 'FamilySize', 'FarePerPerson', 'IsAlone', 'AgeGroup'],
      dtype='object')

## 2. Feature and Target Separation

The target variable is `Survived`, which indicates whether a passenger survived.

All other relevant columns are used as input features.

In [24]:
X = df.drop("Survived", axis=1)

y = df["Survived"]

## 3. Train-Test Split

The dataset is divided into two parts:

- 80% training data
- 20% testing data

The training set is used to train the models, while the testing set is used to evaluate their performance on unseen data.

In [29]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)
print("Training features:", X_train.shape)
print("Testing features:", X_test.shape)

print("Training target:", y_train.shape)
print("Testing target:", y_test.shape)

Training features: (660, 12)
Testing features: (165, 12)
Training target: (660,)
Testing target: (165,)


In [30]:
df.select_dtypes(include=["object", "category"]).columns

Index([], dtype='object')

In [31]:
df.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,FamilySize,FarePerPerson,IsAlone,AgeGroup,Embarked_Q,Embarked_S
0,0,3,1,-0.640556,1,0,-0.485389,2,-0.242695,0,1,False,True
1,1,1,0,0.932531,1,0,0.797171,2,0.398586,0,0,False,False
2,1,3,0,-0.247284,0,0,-0.471869,1,-0.471869,1,1,False,True
3,1,1,0,0.637577,1,0,0.432967,2,0.216484,0,0,False,True
4,0,3,1,0.637577,0,0,-0.469366,1,-0.469366,1,0,False,True


## 4. Logistic Regression

Logistic Regression is a linear classification algorithm used to predict binary outcomes.

In this project, it predicts:

- 0 = Did not survive
- 1 = Survived

In [32]:
logistic_model = LogisticRegression(max_iter=1000)

logistic_model.fit(X_train, y_train)
y_pred_lr = logistic_model.predict(X_test)

In [33]:
lr_accuracy = accuracy_score(y_test, y_pred_lr)
lr_precision = precision_score(y_test, y_pred_lr)
lr_recall = recall_score(y_test, y_pred_lr)
lr_f1 = f1_score(y_test, y_pred_lr)

print("Logistic Regression")
print("Accuracy:", lr_accuracy)
print("Precision:", lr_precision)
print("Recall:", lr_recall)
print("F1 Score:", lr_f1)

Logistic Regression
Accuracy: 0.793939393939394
Precision: 0.7457627118644068
Recall: 0.6984126984126984
F1 Score: 0.7213114754098361


## 5. Decision Tree Classifier

A Decision Tree makes predictions by splitting data based on feature values.

It creates a tree-like structure of decision rules to classify observations.

In [34]:
decision_tree = DecisionTreeClassifier(
    random_state=42
)

decision_tree.fit(X_train, y_train)
y_pred_dt = decision_tree.predict(X_test)

In [35]:
dt_accuracy = accuracy_score(y_test, y_pred_dt)
dt_precision = precision_score(y_test, y_pred_dt)
dt_recall = recall_score(y_test, y_pred_dt)
dt_f1 = f1_score(y_test, y_pred_dt)

print("Decision Tree")
print("Accuracy:", dt_accuracy)
print("Precision:", dt_precision)
print("Recall:", dt_recall)
print("F1 Score:", dt_f1)

Decision Tree
Accuracy: 0.793939393939394
Precision: 0.7301587301587301
Recall: 0.7301587301587301
F1 Score: 0.7301587301587301


## 6. Random Forest Classifier

Random Forest is an ensemble learning algorithm that combines multiple decision trees to make predictions.

It generally provides better generalization than a single decision tree and can reduce overfitting.

In [37]:
random_forest = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

random_forest.fit(X_train, y_train)
y_pred_rf = random_forest.predict(X_test)

In [38]:
rf_accuracy = accuracy_score(y_test, y_pred_rf)
rf_precision = precision_score(y_test, y_pred_rf)
rf_recall = recall_score(y_test, y_pred_rf)
rf_f1 = f1_score(y_test, y_pred_rf)

print("Random Forest")
print("Accuracy:", rf_accuracy)
print("Precision:", rf_precision)
print("Recall:", rf_recall)
print("F1 Score:", rf_f1)

Random Forest
Accuracy: 0.806060606060606
Precision: 0.7719298245614035
Recall: 0.6984126984126984
F1 Score: 0.7333333333333333


In [40]:
comparison = pd.DataFrame({
    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "Random Forest"
    ],

    "Accuracy": [
        lr_accuracy,
        dt_accuracy,
        rf_accuracy
    ],

    "Precision": [
        lr_precision,
        dt_precision,
        rf_precision
    ],

    "Recall": [
        lr_recall,
        dt_recall,
        rf_recall
    ],

    "F1 Score": [
        lr_f1,
        dt_f1,
        rf_f1
    ]
})
#highlight the highest score in each metric
comparison.round(4).style.highlight_max(
    subset=["Accuracy", "Precision", "Recall", "F1 Score"],
    axis=0
)

,Model,Accuracy,Precision,Recall,F1 Score
0,Logistic Regression,0.793900,0.745800,0.698400,0.721300
1,Decision Tree,0.793900,0.730200,0.730200,0.730200
2,Random Forest,0.806100,0.771900,0.698400,0.733300


In [41]:
best_model = comparison.loc[
    comparison["F1 Score"].idxmax()
]

best_model

,2
Model,Random Forest
Accuracy,0.806061
Precision,0.77193
Recall,0.698413
F1 Score,0.733333


## Best Model Selection

Based on the evaluation results, Random Forest was selected as the best overall-performing model.

Random Forest achieved the highest accuracy (80.61%), precision (77.19%), and F1 score (73.33%) among the three models. Although the Decision Tree achieved a slightly higher recall (73.02%) compared to Random Forest (69.84%), Random Forest provided the strongest overall balance across the evaluation metrics.

The F1 score was also considered important because it provides a balance between precision and recall. Random Forest achieved the highest F1 score of 73.33%, slightly outperforming the Decision Tree (73.02%) and Logistic Regression (72.13%).

Therefore, Random Forest was selected as the best overall model for this classification task based on its overall performance across accuracy, precision, recall, and F1 score.